File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/library/LibraryFactory.java
package ru.kmorozov.gbd.core.logic.library;

import ru.kmorozov.gbd.core.logic.connectors.HttpConnector;
import ru.kmorozov.gbd.core.logic.library.metadata.GoogleBooksMetadata;
import ru.kmorozov.gbd.core.logic.library.metadata.RfbrMetadata;
import ru.kmorozov.gbd.core.logic.library.metadata.ShplMetadata;

import java.util.List;

/**
 * Created by sbt-morozov-kv on 15.11.2016.
 */
public final class LibraryFactory {

    private static final ILibraryMetadata[] METADATA = {GoogleBooksMetadata.GOOGLE_METADATA, ShplMetadata.SHPL_METADATA, RfbrMetadata.RFBR_METADATA};

    private static ILibraryMetadata selectedMetadata;

    private LibraryFactory() {
    }

    public static ILibraryMetadata getMetadata(String bookId) {
        for (ILibraryMetadata _metadata : LibraryFactory.METADATA)
            if (_metadata.isValidId(bookId)) {
                LibraryFactory.selectedMetadata = _metadata;
                return _metadata;
            }

        return null;
    }

    public static boolean isValidId(String bookId) {
        return null != LibraryFactory.getMetadata(bookId);
    }

    public static boolean needSetCookies() {
        return LibraryFactory.selectedMetadata.needSetCookies();
    }

    public static List<HttpConnector> preferredConnectors() {
        return LibraryFactory.selectedMetadata.preferredConnectors();
    }
}

File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/library/LibraryFactory.kt
package ru.kmorozov.gbd.core.logic.library

import ru.kmorozov.gbd.core.logic.library.metadata.ArchiveMetadata.Companion.ARCHIVE_METADATA
import ru.kmorozov.gbd.core.logic.library.metadata.GoogleBooksMetadata.Companion.GOOGLE_METADATA
import ru.kmorozov.gbd.core.logic.library.metadata.RfbrMetadata.Companion.RFBR_METADATA
import ru.kmorozov.gbd.core.logic.library.metadata.ShplMetadata.Companion.SHPL_METADATA
import ru.kmorozov.gbd.core.logic.library.metadata.UnknownMetadata.Companion.UNKNOWN_METADATA

/**
 * Created by sbt-morozov-kv on 15.11.2016.
 */
object LibraryFactory {

    private val METADATA = arrayOf(GOOGLE_METADATA, SHPL_METADATA, RFBR_METADATA, ARCHIVE_METADATA, UNKNOWN_METADATA)

    fun getMetadata(bookId: String): ILibraryMetadata {
        return METADATA.first { it.isValidId(bookId) }
    }

    fun isValidId(bookId: String): Boolean {
        return !getMetadata(bookId).equals(UNKNOWN_METADATA)
    }
}
